### ISSUE TITLE
Output After converting weights file into IR is weird

### ISSUE THREAD

**Issue by bhargavravat** [AlexeyAB/darknet#3040]
Hi @AlexeyAB :

I trained yolo V3 using your repo with the below cfg file. 
```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 500200
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=128
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=256
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=512
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

######################

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 6,7,8
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1


[route]
layers = -4

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 61



[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 3,4,5
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1



[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 36



[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=18
activation=linear


[yolo]
mask = 0,1,2
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=1
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1
```

Now , when I run darknet detctor demo it is working fine and obtaining the expected output.  However after converting the weights in to IR using intel openvino the results are extremely weird. I guess I made some mistake in conversion. 
 Step1 : weights to pb conversion 
(using https://github.com/mystic123/tensorflow-yolo-v3 ) 

_`python3 /home/paperspace/Desktop/tensorflow-yolo-v3/convert_weights_pb.py --class_names /home/paperspace/Downloads/Dataset/metadata/person.names --data_format NHWC --weights_file /home/paperspace/Downloads/Dataset/metadata/yolo_backup/yolov3_edited_14700.weights --output_graph /home/paperspace/Desktop/yolo_edited_test_14700.pb --size 416`_

And Now PB to XML 

_`python3 /opt/intel/computer_vision_sdk_2018.5.455/deployment_tools/model_optimizer/mo_tf.py --input_model /home/paperspace/Desktop/yolo_edited_test_14700.pb --tensorflow_object_detection_api_pipeline_config /home/paperspace/Downloads/Dataset/metadata/yolov3_edited.cfg --tensorflow_use_custom_operations_config /opt/intel/computer_vision_sdk_2018.5.455/deployment_tools/model_optimizer/extensions/front/tf/yolo_v3_edited.json --reverse_input_channels --input_shape [1,416,416,3] --data_type FP16 --output_dir /home/paperspace/Desktop/14700/_
`

**Can you suggest the head way ? Here is the output.** 

![1556200491_4418402_S_u_p_p_o_r_t_](https://user-images.githubusercontent.com/11025174/56741489-95f00f00-6790-11e9-87d4-35acadb6f585.png)
![1556200493_9478016_A_P_T_](https://user-images.githubusercontent.com/11025174/56741504-9b4d5980-6790-11e9-9cd5-38aaf52feee3.png)
![1556200495_1629899_A_P_T_ _T_o_w_e_r_](https://user-images.githubusercontent.com/11025174/56741510-9dafb380-6790-11e9-80ca-c92c28297f67.png)
![1556200504_3858953_D_o_o_r_ _C_o_r_r_i_d_o_r_](https://user-images.githubusercontent.com/11025174/56741519-a1433a80-6790-11e9-8c1d-4461a7ccd1ea.png)
![1556200533_7848542_I_o_T_](https://user-images.githubusercontent.com/11025174/56741524-a43e2b00-6790-11e9-8862-602776ec3dae.png)

**Comment by AlexeyAB** [comment]
@bhargavravat Hi,

* Can you successfully convert default model? https://pjreddie.com/media/files/yolov3.weights and https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3.cfg

* Try to set `batch=1 subdivisions=1` before convertion to pb-model

**Comment by bhargavravat** [comment]
hI @AlexeyAB :  I already converted weights https://pjreddie.com/media/files/yolov3.weights with default cfg i.e. https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3.cfg . And its working fine . The issue arises when I am using custom weights .

**Comment by AlexeyAB** [comment]
@bhargavravat 

Change these lines:
```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
```

to these
```
[net]
# Testing
batch=1
subdivisions=1
```

**Comment by bhargavravat** [comment]
> @bhargavravat
> 
> Change these lines:
> 
> ```
> [net]
> # Testing
> #batch=1
> #subdivisions=1
> # Training
> batch=64
> subdivisions=16
> ```
> to these
> 
> ```
> [net]
> # Testing
> batch=1
> subdivisions=1
> ```

@AlexeyAB : 

Changed the cfg file as per your suggestion still there is no change in the output.!

**Comment by bhargavravat** [comment]
@AlexeyAB :
While converting IR files , total 3 files are generated . 
1. .xml , .bin , .mapping 

Here is the comparison of .mapping file . Left sided file is with default configuration (https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3.cfg) and original weights. 
(https://pjreddie.com/media/files/yolov3.weights)

![image](https://user-images.githubusercontent.com/11025174/56787459-33435580-681a-11e9-9141-ab1dcd387b5b.png)

**Comment by AlexeyAB** [comment]
@bhargavravat 

Try to convert
* without `--reverse_input_channels --input_shape [1,416,416,3] --data_type FP16 --output_dir /home/paperspace/Desktop/14700/_`
* and without ` --tensorflow_object_detection_api_pipeline_config /home/paperspace/Downloads/Dataset/metadata/yolov3_edited.cfg`

Does it help?


`_python3 /opt/intel/computer_vision_sdk_2018.5.455/deployment_tools/model_optimizer/mo_tf.py --input_model /home/paperspace/Desktop/yolo_edited_test_14700.pb --tensorflow_use_custom_operations_config /opt/intel/computer_vision_sdk_2018.5.455/deployment_tools/model_optimizer/extensions/front/tf/yolo_v3_edited.json`

-----

If it doesn't help, try to ask your question 
* here: https://software.intel.com/en-us/forums/computer-vision
* and here: https://github.com/mystic123/tensorflow-yolo-v3/issues

**Comment by bhargavravat** [comment]
@AlexeyAB :

Can you help me understand this json file 

`[ { "id": "TFYOLOV3", "match_kind": "general", "custom_attributes": { "classes": 1, "coords": 4, "num": 9, "mask": [0, 1, 2], "entry_points": ["detector/yolo-v3/Reshape", "detector/yolo-v3/Reshape_4", "detector/yolo-v3/Reshape_8"] } } ]`

I have changed the value of classes from 80 to 1 . 
What are the other parameters?

**Comment by AlexeyAB** [comment]
@bhargavravat Hi,

> [ { "id": "TFYOLOV3", "match_kind": "general", "custom_attributes": { "classes": 1, "coords": 4, "num": 9, "mask": [0, 1, 2], "entry_points": ["detector/yolo-v3/Reshape", "detector/yolo-v3/Reshape_4", "detector/yolo-v3/Reshape_8"] } } ]

Yes, you should use classes=1 if you trained only for 1 class.
It seems it uses 3 different reshape-layers Reshape/Reshape_4/Reshape_8 (like 3 different yolo-layers in yolov3.cfg). 
But why do the use only masks=0,1,2 - I don't know.

----

Also you can try use your `yolov3-obj.cfg` and `yolov3-obj.weights` without converting
* just to build the latest OpenCV with OpenVINO: https://github.com/opencv/opencv/wiki/Intel's-Deep-Learning-Inference-Engine-backend
* and use OpenCV-dnn with `net.setPreferableBackend(DNN_BACKEND_INFERENCE_ENGINE);` and `net.setPreferableTarget(DNN_TARGET_MYRIAD);`
* Use this example: 
    * c++: https://github.com/opencv/opencv/blob/master/samples/dnn/object_detection.cpp#L192-L221
    * Python: https://github.com/opencv/opencv/blob/master/samples/dnn/object_detection.py#L129-L150

And look at these samples: https://www.learnopencv.com/using-openvino-with-opencv/

**Comment by bhargavravat** [comment]
@AlexeyAB :

I resolved the issue. 

Actual bottle neck was Intel openvino 2018 R5 Version, I upgraded the same to 2019 R1 and things are working smooth now.

Thank you so much for your inputs and kind support!!!

**Comment by andeyeluguo** [comment]
@bhargavravat  which repository did you use to transfrom the darknet ? I am facing the same problems as you, but I use 2020 version.

**Comment by GotG** [comment]
> 
> 
> @bhargavravat which repository did you use to transfrom the darknet ? I am facing the same problems as you, but I use 2020 version.

I am using 2020.1 version and I have multiple boxes erroneous and continuous detections for the converted .xml tiny yolov3 file with custom weights.
did you find a solution?
